File: cses/03_dynamic_programming/1748_increasing-subsequence-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    a = data[1:1 + n]

    MOD = 1000000007

    # Nén tọa độ: gán rank 1..m cho các giá trị phân biệt, tính sẵn rank cho từng phần tử
    # để tránh tra cứu dict trong vòng lặp chính.
    ints = list(map(int, a))
    sorted_vals = sorted(set(ints))
    rank = {v: i + 1 for i, v in enumerate(sorted_vals)}
    m = len(sorted_vals)
    ranks = [rank[v] for v in ints]

    # Fenwick tree (BIT) lưu tổng dp theo rank giá trị.
    # dp luôn được rút gọn < MOD nên mỗi ô tree bị cộng tối đa n lần, giá trị < n*MOD ~ 2e14,
    # vẫn nằm trong phạm vi số nguyên máy (< 2^63). Nhờ đó bỏ được phép mod trong vòng update,
    # giảm hằng số thời gian mà không sợ tràn (mọi phép chỉ là cộng nên lấy mod ở cuối là đủ).
    tree = [0] * (m + 1)

    total = 0
    # Duyệt trái sang phải: khi xử lý phần tử i thì BIT chỉ chứa các j < i.
    for r in ranks:
        # Truy vấn prefix sum trên [1, r-1] -> tổng dp của các phần tử trước có giá trị < v
        # (dừng ở r-1 để bảo đảm tăng NGẶT).
        i = r - 1
        s = 0
        while i > 0:
            s += tree[i]
            i -= i & (-i)
        # dp[i] = 1 + tổng dp[j] thỏa điều kiện; số 1 ứng với dãy chỉ gồm riêng phần tử này.
        dp = (s + 1) % MOD
        total += dp
        # Cập nhật: cộng dp vào vị trí rank r (không rút gọn tree, giá trị vẫn bị chặn an toàn).
        i = r
        while i <= m:
            tree[i] += dp
            i += i & (-i)

    sys.stdout.write(str(total % MOD))


main()

File: cses/03_dynamic_programming/1748_increasing-subsequence-ii/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// Fenwick tree (BIT) lưu tổng dp theo rank giá trị, hỗ trợ prefix sum có lấy mod.
class FenwickTree {
public:
    explicit FenwickTree(int size) : tree_(size + 1, 0) {}

    // Cộng value vào vị trí index rồi lan lên các nút cha; giữ mỗi ô trong [0, MOD).
    void add(int index, int value) {
        while (index < static_cast<int>(tree_.size())) {
            tree_[index] += value;
            if (tree_[index] >= MOD) {
                tree_[index] -= MOD;
            }
            index += index & -index;
        }
    }

    // Tổng dp trên prefix [1, index] theo giá trị, lấy modulo MOD.
    int prefix_sum(int index) const {
        long long result = 0;
        while (index > 0) {
            result += tree_[index];
            index -= index & -index;
        }
        return static_cast<int>(result % MOD);
    }

private:
    static constexpr int MOD = 1'000'000'007;
    vector<int> tree_;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    constexpr int MOD = 1'000'000'007;

    int n;
    cin >> n;
    vector<long long> values(n);
    for (long long& value : values) {
        cin >> value;
    }

    // Nén tọa độ: các giá trị phân biệt sau khi sort sẽ được ánh xạ sang rank 1..m.
    vector<long long> compressed = values;
    sort(compressed.begin(), compressed.end());
    compressed.erase(unique(compressed.begin(), compressed.end()), compressed.end());

    FenwickTree fenwick(static_cast<int>(compressed.size()));
    long long answer = 0;
    // Duyệt từ trái sang phải nên BIT chỉ chứa các j < i, bảo đảm điều kiện j < i.
    for (const long long value : values) {
        const int rank = static_cast<int>(
                             lower_bound(compressed.begin(), compressed.end(), value) - compressed.begin()) +
                         1;

        // dp[i] = 1 + tổng dp[j] với a[j] < a[i]; truy vấn tới rank - 1 để bảo đảm
        // tăng NGẶT (các giá trị bằng nhau không được nối vào nhau).
        const int ending_here = (fenwick.prefix_sum(rank - 1) + 1LL) % MOD;
        answer += ending_here;
        if (answer >= MOD) {
            answer -= MOD;
        }
        // Cộng dp vừa tính vào BIT tại vị trí rank của giá trị hiện tại.
        fenwick.add(rank, ending_here);
    }

    cout << answer << '\n';
    return 0;
}
